designdocs/CAPACITY.md

```diff
@@ -0,0 +1,148 @@
+
+Feeless transactions are important in reaching mass adoption as it removes the overhead costs of transactions for app developers to acquire a far reaching user base.
+
+**What is Capacity?**
+
+Capacity is a non-transferable resource that is associated to an MSA account. You can use capacity to pay for transactions and it replenishes after a certain period which can be used to continue to pay for new transactions. More details on how capacity replenishes below.
+
+## Proposal
+
+MRC has two types of account systems—a MSA account and a token account. You can acquire Capacity by staking your own tokens to your MSA account or when others stake their tokens to your MSA. The  amount staked is reserved from the your account free balance which is refunded after a thaw period. Once staked,  Capacity is granted at a 1 to 1 conversion.  Upon receiving Capacity, it can then be used to pay for transactions. 
+
+There is a minimum required for staking that will be detailed further in a supplemental design that goes into details of staking and rewards.
+
+**Below** **are transactions that can be paid with Capacity:**
+
+- Create an MSA
+- Add a key to an MSA
+- Delegate permissions to another MSA
+- Update Delegate Permissions
+- Send a message
+- Send a batch message
+
+As mentioned above, Capacity is non-transferable and at a minimum implements the following interface.
+
+```rust
+traits Nontransferable {
+   type Balance = Balance;
+
+   fn available(msa_id: MessageSenderId) -> Result<Balance, DispatchError>;
+   fn reduce_available(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, DispatchError>;
+   fn increase_available(msa_id: MessageSenderId, amount: Balance) -> Result<Balance, DispatchError>;
+ }
+
+```
+
+We can represent Capacity as a mapping of an MsaId to information containing the total balance available to an MSA.
+
+```rust
+#[pallet::storage]
+pub type CapacityOf<T: Config> = StorageMap<_, Blake2_128Concat, MessageSenderId, Details<T::Balance>>;
+
+pub struct CapacityDetails {
+	pub used: Balance,
+	pub total_available: Balance,
+
+  // these to be discussed below
+  pub last_replenished_epoch: u32,
+  pub initial_epoch: u32
+}
+```
+
+And we can store the a token account staking to MSA by mapping.
+
+```rust
+#[pallet::storage]
+pub type Staking<T: Config> = StorageMap<_, Blake2_128Concat, T::AccountId, StakingDetails<T::Balance>>;
+
+pub struct StakingDetails<Balance> {
+	pub msa_id: MessageSenderId,
+	pub total_reserved: Balance,
+}
+```
+
+To give the limitations of what you can do with Capacity you are limited to staking and un-staking Capacity for token. Note that these interface can change as we introduce stash and controller accounts using Proxy pallet.
+
+```rust
+pub fn stake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+
+pub fn unstake(origin: OriginFor<T>, account: MessageSenderId, amount: BalanceOf<T>)
+```
```

Comment:
Should there be an initial period after staking such that only after that period has passed the unstaking will be allowed.
But I see this might be replaced with proxy. Is there a scenario where one can increase capacity for a period of epochs as for example when users will expect high volume and tone it down later.

Reply:
Yes. We are leaving the thaw system for https://github.com/LibertyDSNP/mrc/issues/40

Reply:
I was going to ask about a cooling off period for staking/unstaking too.  A thaw system should be noted in this doc and in the Context and Scope it should be stated that it will not be specified in this document.